include/swift/Syntax/SyntaxData.h

```diff
@@ -51,156 +51,154 @@
+  /// RawSyntax's \c SyntaxArena outlives this \c SyntaxDataRef.
+  ///
+  /// In \c SyntaxData, the backing \c SyntaxArena is retained via the \c Arena
+  /// property, lifiting the responsibility to guarantee the \c RawSyntax node
+  /// stays alive from the user.
   const AbsoluteRawSyntax AbsoluteRaw;
-  RC<RefCountedBox<SyntaxData>> Parent;
 
-  /// If this node is the root of a Syntax tree (i.e. \c Parent is \c nullptr ),
-  /// the arena in which this node's \c RawSyntax node has been allocated.
-  /// This keeps this \c RawSyntax nodes referenced by this tree alive.
-  RC<SyntaxArena> Arena;
-
-  /// Create a intermediate node with a parent.
-  SyntaxData(AbsoluteRawSyntax AbsoluteRaw,
-             const RC<RefCountedBox<SyntaxData>> &Parent)
-      : AbsoluteRaw(AbsoluteRaw), Parent(Parent), Arena(nullptr) {}
-
-  /// Create a new root node
-  SyntaxData(AbsoluteRawSyntax AbsoluteRaw)
-      : AbsoluteRaw(AbsoluteRaw), Parent(nullptr),
-        Arena(AbsoluteRaw.getRaw()->getArena()) {}
-
-public:
-  /// With a new \c RawSyntax node, create a new node from this one and
-  /// recursively rebuild the parental chain up to the root.
-  SyntaxData replacingSelf(const RawSyntax *NewRaw) const;
-
-  /// Replace a child in the raw syntax and recursively rebuild the
-  /// parental chain up to the root.
-  template <typename CursorType>
-  SyntaxData replacingChild(const RawSyntax *RawChild,
-                            CursorType ChildCursor) const {
-    auto NewRaw = AbsoluteRaw.getRaw()->replacingChild(ChildCursor, RawChild);
-    return replacingSelf(NewRaw);
+  /// The parent of this node. The parent is *not* retained by the \c
+  /// SyntaxDataRef. It is the user's responsibility to ensure that the parent
+  /// always outlives the child.
+  ///
+  /// If this is a \c SyntaxData object, \c Parent is guaranteed to also be a
+  /// \c SyntaxData object, which is manually retained when \c this is created
+  /// and released when \c this is destroyed.
+  const SyntaxDataRef *Parent;
+
+  /// Creates an empty \c SyntaxDataRef variable to which values can later be
+  /// stored.
+  SyntaxDataRef() : AbsoluteRaw() {}
+
+  /// Creates a \c SyntaxDataRef. \p AbsoluteRaw must not be null. If \p Parent
+  /// is a \c nullptr, this node represents the root of a tree.
+  SyntaxDataRef(const AbsoluteRawSyntax &AbsoluteRaw,
+                const SyntaxDataRef *Parent)
+      : AbsoluteRaw(AbsoluteRaw), Parent(Parent) {
+    assert(!AbsoluteRaw.isNull() &&
+           "SyntaxDataRef must always have a value. Use default constructor to "
+           "create a SyntaxDataRef to be populated later.");
   }
 
-  /// Get the node immediately before this current node that does contain a
-  /// non-missing token. Return \c None if we cannot find such node.
-  Optional<SyntaxData> getPreviousNode() const;
+#ifndef NDEBUG
+  virtual bool isRef() const { return true; }
+  virtual ~SyntaxDataRef() {}
```

Comment:
Maybe a dumb C++ question. Could you explain why this is in `#ifndef`?

Reply:
Because it’s `#ifndef NDEBUG` everywhere else as well. I always assumed this was to make the default a debug build.
If you use `#ifdef DEBUG`, the default is a non-debug build unless you define `DEBUG`.

Reply:
I mean why the destructor must be in the `#ifndef`?